api-catalog-services/src/main/java/com/ca/mfaas/apicatalog/services/cached/CacheRefreshService.java

```diff
@@ -126,20 +129,26 @@ public void refreshCacheFromDiscovery() {
 
     /**
      * Get this instance service details and check if it should be processed
-     *
-     * @param containersUpdated  which containers were updated
+     * @param containersUpdated which containers were updated
+     * @param cachedServices existing services
      * @param deltaFromDiscovery changed service instances
-     * @param instance           this instance
+     * @param instance this instance
      */
-    private void processServiceInstance(Set<String> containersUpdated,
-                                        Applications deltaFromDiscovery,
-                                        InstanceInfo instance) {
-        deltaFromDiscovery.getRegisteredApplications()
-            .stream()
-            .filter(service -> service.getName().equalsIgnoreCase(instance.getAppName()))
-            .findFirst()
-            .ifPresent(application -> processInstance(containersUpdated, instance, application));
+    private void processServiceInstance(Set<String> containersUpdated, Applications cachedServices,
+                                        Applications deltaFromDiscovery, InstanceInfo instance) {
+        Application application = null;
+        // Get the application which this instance belongs to
+        if (cachedServices != null && cachedServices.getRegisteredApplications() != null) {
+            application = cachedServices.getRegisteredApplications().stream()
+                .filter(service -> service.getName().equalsIgnoreCase(instance.getAppName())).findFirst().orElse(null);
+        }
+        // if its new then it will only be in the delta
+        if (application == null || application.getInstances().isEmpty()) {
+            application = deltaFromDiscovery.getRegisteredApplications().stream()
+                .filter(service -> service.getName().equalsIgnoreCase(instance.getAppName())).findFirst().orElse(null);
+        }
 
+        processInstance(containersUpdated, instance, application);
```

Comment:
The `application` could be a null here, because of `.orElse(null)`, is in it? If so, wouldn't be a better to check on null here or inside `processInstance` method, but before this `application.removeInstance(instance);`?

Reply:
Yes, you're right. After getting this implementation from old commits, I started to think that can it be null?
After that, I saw that it can't be null. Because `processServiceInstance` is called in iterations and this iteration was getting from deltaFromDiscovery.getRegisteredApplications().
    - But currently, I think it might be covered against NPException even it can't be null. I will do it.
Thanks!